### PR TITLE
Log warnings for strange snapshot status updates

### LINKED ISSUE #1715: Log when "something wierd" happens in the ensemble
A first step towards a state-machine in the ensemble is to print / log when the ensemble gets states that are out of order / states that are not expected.

From this we can decide spawn more issues 

Spawned from: #1691

Depends on: https://github.com/equinor/ert/issues/1717

### PR DESCRIPTION
**Pull request by DanSava**
**Issue**
Resolves #1715

**Approach**
Add a list of approved state transitions.
When the snapshot state is updated check if the transition is approved if not log a warning.

### CODE REVIEW

**Comment by jondequinor** [comment]
I suggest going full state machine. I've done this spartan approach a couple of times, and I always regret it.

**Comment by DanSava** [comment]
> I suggest going full state machine. I've done this spartan approach a couple of times, and I always regret it.

Not really sure it is worth adding all the logic needed for a full state machine and testing for it just now.

**Comment by DanSava** [comment]
But let's also explore what a minimal state machine approach would look like.

The allowed states and transitions for an ensemble would be:

Screenshot from 2021-06-02 13-34-23

**Comment by jondequinor** [comment]
> Not really sure it is worth adding all the logic needed for a full state machine and testing for it just now.

I think the premise that the cost of doing it "fully" now is less than deploying a lesser solution now and fixing it later—is false.

**Comment by jondequinor** [comment]
> Anything else than what is above I would consider a strange state transition right?

I think you're missing Starting -> Failed? Maybe Unknown -> Cancel?

But do we have Unknown for the ensemble?

**Comment by DanSava** [comment]
> But do we have Unknown for the ensemble?

We will after this PR :wink:

**Comment by DanSava** [comment]
> I think you're missing Starting -> Failed? Maybe Unknown -> Cancel?

You are right I forgot to draw the line from started to Failed.

But is it possible to go from unknown directly to cancelled without first being started?

**Comment by jondequinor** [comment]
> > I think you're missing Starting -> Failed? Maybe Unknown -> Cancel?
> 
> You are right I forgot to draw the line from started to Failed.
> 
> But is it possible to go from unknown directly to cancelled without first being started?

Not in the legacy ensemble. Could you say whether or not it will be possible for Prefect?

**Comment by DanSava** [comment]
Not really possible anywhere the ensemble snapshot is created with Started status :worried: from the get go

File: ert_shared/ensemble_evaluator/entity/ensemble_base.py [hunk 1]
```diff
@@ -12,19 +14,76 @@
...
-    JOB_STATE_START,
-    REALIZATION_STATE_WAITING,
-    STEP_STATE_UNKNOWN,
-)
+import ert_shared.status.entity.state as state
+
+logger = logging.getLogger(__name__)
+
+
+class _EnsembleStateTracker:
```
Comment (joakim-hove) [inline ert_shared/ensemble_evaluator/entity/ensemble_base.py:22]: This looks more like a run state than an ensemble state - an ensemble is a meaningful object without concept of simulation and running.

Comment (jondequinor) [reply ert_shared/ensemble_evaluator/entity/ensemble_base.py:22]: Yeah, agree—it should really be response or realisation.

Comment (jondequinor) [reply ert_shared/ensemble_evaluator/entity/ensemble_base.py:22]: … but I don't think you should change this Dan. Even though the naming is a bit off, it's what we currently got. I can create an issue that could track implementation of proper naming, reflection on concepts, etc.

**Review by frode-aarstad — APPROVED** [review]

**Comment by frode-aarstad** [comment]
The current implementation looks good. The discussion regarding making it a statemachine or not using ensemble as a name is probably another discussion / issue

**Pull request merged**